### PR TITLE
[X86] Fold `(icmp ult (add x,-C),2)` -> `(or (icmp eq X,C), (icmp eq X,C+1))` for Vectors

### LINKED ISSUE #66479: [X86] Poor codegen for SSE ((x == 1) || (x == 2))
https://godbolt.org/z/KT5fxKjbG
```c
#include <x86intrin.h>
__m128i cmp1OR2_epi32(__m128i x) {
    __m128i is1 = _mm_cmpeq_epi32(x, _mm_set1_epi32(1));
    __m128i is2 = _mm_cmpeq_epi32(x, _mm_set1_epi32(2));
    return _mm_or_si128(is1, is2);
}
```
I used 1 and 2, but it occurs with any 2 sequential constants.

Due to an InstCombine fold, instead of just comparing the 2 sequential values and ORing the result, we end up with:
```ll
define <2 x i64> @cmp1OR2_epi32(<2 x i64> noundef %x) {
entry:
  %0 = bitcast <2 x i64> %x to <4 x i32>
  %1 = add <4 x i32> %0, <i32 -1, i32 -1, i32 -1, i32 -1>
  %or.i89 = icmp ult <4 x i32> %1, <i32 2, i32 2, i32 2, i32 2>
  %or.i8 = sext <4 x i1> %or.i89 to <4 x i32>
  %or.i = bitcast <4 x i32> %or.i8 to <2 x i64>
  ret <2 x i64> %or.i
}
```
```asm
cmp1OR2_epi32:
  pcmpeqd %xmm1, %xmm1
  paddd %xmm1, %xmm0
  movdqa .LCPI2_0(%rip), %xmm1 # xmm1 = [1,1,1,1]
  pminud %xmm0, %xmm1
  pcmpeqd %xmm1, %xmm0
  retq
```
I could /almost/ accept this if it meant we reduced constant loads, but on pre-SSE4  targets we end up with:
```asm
cmp1OR2_epi32:
  pcmpeqd %xmm1, %xmm1
  paddd %xmm0, %xmm1
  pxor .LCPI2_0(%rip), %xmm1
  movdqa .LCPI2_1(%rip), %xmm0 # xmm0 = [2147483650,2147483650,2147483650,2147483650]
  pcmpgtd %xmm1, %xmm0
  retq
```
Untangling the icmp_ult(add(x,-1),2) (or the general solution) shouldn't be too difficult.

### PR DESCRIPTION
**Pull request by goldsteinn**
This is undoing a middle-end transform which does the opposite. Since
X86 doesn't have unsigned vector comparison instructions pre-AVX512,
the simplified form gets worse codegen.

Fixes https://github.com/llvm/llvm-project/issues/66479

Proofs: https://alive2.llvm.org/ce/z/UCz3wt

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -64,6 +64,7 @@
 #include <algorithm>
 #include <bitset>
 #include <cctype>
+#include <llvm-19/llvm/CodeGen/ISDOpcodes.h>
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:67]: Remove this?

Out of interest - what tool keeps adding these ?

Comment (goldsteinn) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:67]: Ill keep better watch for these in the future.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -53408,6 +53409,64 @@ static SDValue combineSetCC(SDNode *N, SelectionDAG &DAG,
...
+      // See which form we have depending on the constant/condition.
+      SDValue C0 = SDValue();
+      SDValue C1 = SDValue();
+
+      // If we had `(add x, -1)` and can lower with `umin`, don't transform as
+      // we will end up generating an additional constant. Keeping in the
+      // current form has a slight latency cost, but it probably worth saving a
+      // constant.
+      if (ISD::isConstantSplatVectorAllOnes(AddC.getNode()) &&
+          DAG.getTargetLoweringInfo().isOperationLegal(ISD::UMIN, OpVT)) {
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:53441]: Yeah, seems this isn't beneficial with `AVX1`. End up with more `vinsertf128`. Ill just drop transform for 256-bit + AVX

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request closed without merging**